src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/TextDiffer.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    //
+    // This class implements the linear space variation of the difference algorithm described in
+    // "An O(ND) Difference Algorithm and its Variations" by Eugene W. Myers
+    //
+    // Note: Some variable names in this class are not be fully compliant with the C# naming guidelines
+    // in the interest of using the same terminology discussed in the paper for ease of understanding.
+    // 
+    internal abstract class TextDiffer
+    {
+        protected abstract int OldTextLength { get; }
+
+        protected abstract int NewTextLength { get; }
+
+        protected abstract bool ContentEquals(int oldTextIndex, int newTextIndex);
+
+        protected IReadOnlyList<DiffEdit> ComputeDiff()
+        {
+            var edits = new List<DiffEdit>();
+
+            // Initialize the vectors to use for forward and reverse searches.
+            var MAX = NewTextLength + OldTextLength;
+            var Vf = new int[(2 * MAX) + 1];
+            var Vr = new int[(2 * MAX) + 1];
+
+            ComputeDiffRecursive(edits, 0, OldTextLength, 0, NewTextLength, Vf, Vr);
+
+            return edits;
+        }
+
+        private void ComputeDiffRecursive(List<DiffEdit> edits, int lowA, int highA, int lowB, int highB, int[] Vf, int[] Vr)
```

Comment:
![image](https://media0.giphy.com/media/2rqEdFfkMzXmo/giphy.gif?cid=ecf05e47dd541c3a8583982c0538b71e55677898afcc9b0b&rid=giphy.gif)

Reply:
Haha yeah it definitely took multiple days to understand the whole thing. The algorithm itself is really elegant in theory but the implementation gets very tedious.